### ISSUE TITLE
It seems that the valid sets are not used at all?

### ISSUE THREAD

**Issue by yuchenlin** [thunlp/OpenKE#38]
Hi there,

I suppose that valid set is typically used to do the early stop in the training phase. I understand that running a test on a valid set after each epoch is very time-consuming, while it is better to have that option. 

Also, the other important function of valid sets, I think, is to select the hyper-parameters. However, the test phase seems to be only able to run on the test set. Are the valid sets used in this framework at all?

**Comment by pmitra01** [comment]
It seems that they are used in the Triple Classification task, (See, for instance Config.py, where the tensors for validation set batches are initialised; and are also tested, under the test method). However, I have exclusively stuck to link prediction, and have not implemented  OpenKE's Triple Classification,and hence can not vouch for the results.)

In the link prediction task, the more I think about it, the more it seems that early stopping based simply on the loss which is minimized to learn to score better based on positive examples (+1) and a a few negatively sampled examples (-1) may not be ideal, as the problem at hand forms more of an information retrieval task. Here we are more concerned with the top n candidates in the position of head or tail of a given triple, measured by Hits@n, or MRR, and not simply the precision or recall. (hits@1 is the the conventional precision metric as we know it. Perhaps there is a way to choose a different metric to optimise on, on the validation set, but it seems very computationally expensive to carry out a ranking procedure each time on the validation set. 

I may be wrong, and my line of reasoning is more biased towards "why we would not want a validation-based optimisation process",rather than "why we want one." Primarily because I do not recall most seminal works in this field reporting their results on a validation set. It would be great to hear other people's thoughts on this.

**Comment by UditGupta10** [comment]
I think that for early stopping we can have 2 validation batches valid1 and valid2. The threshold is determined by valid1 and we can check the triple_classification accuracy on valid2. We can early stop if no significant improvement is seen in valid2's triple_classification accuracy. That way we can leave aside the test set for its original purpose.

**Comment by zyksir** [comment]
it seems that the latest version still has this problem. After an update for new version of pytorch, you delete all the code for validation
